pkg/util/testutil/patternscanner.go

```diff
@@ -61,17 +78,40 @@ func (ps *PatternScanner) Write(p []byte) (n int, err error) {
 	// Process all complete lines.
 	for _, line := range lines[:len(lines)-1] {
 		ps.buffers = append(ps.buffers, line) // Save the log line.
-		if !ps.stopped && ps.pattern.MatchString(line) {
-			ps.stopOnce.Do(func() {
-				ps.stopped = true
-				close(ps.DoneChan) // Notify the caller about the match.
-			})
-		}
+
+		// Check if we've met the scanning criteria
+		ps.matchPatterns(line)
 	}
 
 	return len(p), nil
 }
 
+// matchPatterns checks if the current line matches the scanning requirements
+func (ps *PatternScanner) matchPatterns(line string) {
+	switch {
+	// started pattern not found yet, look for it
+	case !ps.started:
+		if ps.startPattern.MatchString(line) {
+			// found start pattern, flip the flag to start looking for finish pattern for following iterations
+			ps.started = true
+
+			if ps.finishPattern == nil {
+				ps.notifyAndStop()
```

Comment:
perhaps `fallthrough` can be better?

Reply:
the `fallthrough` with current cases won't work, because the second case has the opposite condition (finishPattern is provided and we should try and match it.

Reply:
I meant to have

```suggestion
if ps.finishPattern == nil {
    fallthrough
}
```
instead of calling `ps.notifyAndStop()`

Reply:
i don't think you can use fallthrough inside the if statement. it must be the last statement of the block